File: Samples/ADGSample/NativeAdsObjC/ADGNativeAdView/ADGNativeAdView.m
//
//  ADGNativeAdView.m
//  NativeAdsObjC
//
//  Copyright © 2017年 Supership Inc. All rights reserved.
//

#import "ADGNativeAdView.h"
#import <ADG/ADGInformationIconView.h>
#import <ADG/ADGMediaView.h>
#import <ADG/ADGNativeAd.h>

@interface ADGNativeAdView ()

@property (weak, nonatomic) IBOutlet UIImageView *iconImageView;
@property (weak, nonatomic) IBOutlet UILabel *titleLabel;
@property (weak, nonatomic) IBOutlet UILabel *descriptionLabel;
@property (weak, nonatomic) IBOutlet UIView *mediaViewContainerView;
@property (weak, nonatomic) IBOutlet UILabel *sponsoredLabel;
@property (weak, nonatomic) IBOutlet UILabel *ctaLabel;

@end

@implementation ADGNativeAdView

+ (instancetype) view {
    NSString *className = NSStringFromClass([self class]);

    return [[[NSBundle mainBundle] loadNibNamed:className
                                          owner:nil
                                        options:0] firstObject];
}

- (void) awakeFromNib {
    [super awakeFromNib];
    self.layer.borderWidth = 1.0f;
    self.layer.borderColor = UIColor.lightGrayColor.CGColor;

    self.iconImageView.contentMode = UIViewContentModeScaleAspectFit;
    self.iconImageView.clipsToBounds = YES;

    self.ctaLabel.backgroundColor = UIColor.whiteColor;
    self.ctaLabel.clipsToBounds = YES;
    self.ctaLabel.textColor = self.ctaLabel.tintColor;
    self.ctaLabel.layer.borderWidth = 1.0f;
    self.ctaLabel.layer.borderColor = self.ctaLabel.tintColor.CGColor;
    self.ctaLabel.layer.cornerRadius = 5.0f;
}

- (void) apply:(ADGNativeAd *)nativeAd viewController:(UIViewController *)viewController {
    if (!nativeAd) {
        return;
    }
    // タイトル
    self.titleLabel.text = nativeAd.title ? nativeAd.title.text : @"";

    // リード文
    self.descriptionLabel.text = nativeAd.desc ? nativeAd.desc.value : @"";

    // アイコン画像
    if (nativeAd.iconImage.url.length > 0) {
        NSURL *iconImageUrl = [NSURL URLWithString:nativeAd.iconImage.url];
        __weak typeof(self) weakSelf = self;
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            NSData *iconImageData = [NSData dataWithContentsOfURL:iconImageUrl];
            dispatch_async(dispatch_get_main_queue(), ^{
                weakSelf.iconImageView.image = [UIImage imageWithData:iconImageData];
            });
        });
    }

    // メイン画像・動画
    for (UIView *v in self.mediaViewContainerView.subviews) {
        [v removeFromSuperview];
    }

    if (nativeAd.canLoadMedia) {
        CGRect mediaViewFrame = CGRectMake(0, 0,
                                           self.mediaViewContainerView.bounds.size.width,
                                           self.mediaViewContainerView.bounds.size.height);
        ADGMediaView *mediaView = [[ADGMediaView alloc] initWithFrame:mediaViewFrame];
        mediaView.nativeAd = nativeAd;
        mediaView.viewController = viewController;
        [self.mediaViewContainerView addSubview:mediaView];
        [mediaView load];
    }

    // インフォメーションアイコン
    ADGInformationIconView *infoIconView = [[ADGInformationIconView alloc] initWithNativeAd:nativeAd];
    if (infoIconView) {
        [self.mediaViewContainerView addSubview:infoIconView];
        [infoIconView updateFrameFromSuperview:UIRectCornerTopRight];
    }

    // 広告主
    self.sponsoredLabel.text = nativeAd.sponsored.value.length > 0 ?
        [NSString stringWithFormat:@"sponsored by %@", nativeAd.sponsored.value] :
        @"sponsored";

    // CTAボタン
    NSString *ctaText = nativeAd.ctatext.value.length > 0 ?
        nativeAd.ctatext.value : @"詳しくはこちら";
    self.ctaLabel.text = ctaText;

    // クリックイベント
    [nativeAd setTapEvent:self handler:nil];
}

@end

File: Samples/ADGSample/NativeAdsSwift/ADGNativeAdView/ADGNativeAdView.swift
//
//  ADGNativeAdView.swift
//  NativeAdsSwift
//
//  Copyright © 2017年 Supership Inc. All rights reserved.
//
import UIKit

import ADG

class ADGNativeAdView: UIView {
    @IBOutlet var iconImageView: UIImageView! {
        didSet {
            iconImageView.contentMode = .scaleAspectFit
            iconImageView.clipsToBounds = true
        }
    }

    @IBOutlet var titleLabel: UILabel!
    @IBOutlet var descriptionLabel: UILabel!
    @IBOutlet var mediaViewContainerView: UIView!
    @IBOutlet var sponsoredLabel: UILabel!
    @IBOutlet var ctaLabel: UILabel! {
        didSet {
            ctaLabel.backgroundColor = UIColor.white
            ctaLabel.clipsToBounds = true
            ctaLabel.textColor = ctaLabel.tintColor
            ctaLabel.layer.borderWidth = 1.0
            ctaLabel.layer.borderColor = ctaLabel.tintColor.cgColor
            ctaLabel.layer.cornerRadius = 5.0
        }
    }

    static func view() -> ADGNativeAdView {
        let className = String(describing: self)
        return Bundle.main.loadNibNamed(className, owner: nil, options: nil)?
            .first as! ADGNativeAdView
    }

    func apply(nativeAd: ADGNativeAd, viewController: UIViewController) {
        layer.borderWidth = 1.0
        layer.borderColor = UIColor.lightGray.cgColor

        // タイトル
        titleLabel.text = nativeAd.title?.text ?? ""

        // リード文
        descriptionLabel.text = nativeAd.desc?.value ?? ""

        // アイコン画像
        if let urlStr = nativeAd.iconImage?.url,
           let url = URL(string: urlStr)
        {
            DispatchQueue.global().async {
                if let data = try? Data(contentsOf: url) {
                    DispatchQueue.main.async { [weak self] in
                        self?.iconImageView.image = UIImage(data: data)
                    }
                }
            }
        }

        // メイン画像・動画
        mediaViewContainerView.subviews.forEach {
            $0.removeFromSuperview()
        }

        if nativeAd.canLoadMedia {
            let frame = CGRect(x: 0, y: 0,
                               width: mediaViewContainerView.bounds.size.width,
                               height: mediaViewContainerView.bounds.size.height)
            let mediaView = ADGMediaView(frame: frame)
            mediaView.nativeAd = nativeAd
            mediaView.viewController = viewController
            mediaViewContainerView.addSubview(mediaView)
            mediaView.load()
        }

        // インフォメーションアイコン
        let infoIconView = ADGInformationIconView(nativeAd: nativeAd)
        mediaViewContainerView.addSubview(infoIconView)
        infoIconView.updateFrame(fromSuperview: .topRight)

        // 広告主
        if let sponsoredText = nativeAd.sponsored?.value,
           !sponsoredText.isEmpty
        {
            sponsoredLabel.text = "sponsored by \(sponsoredText)"
        } else {
            sponsoredLabel.text = "sponsored"
        }

        // CTAボタン
        if let ctaText = nativeAd.ctatext?.value {
            ctaLabel.text = ctaText.isEmpty ? ctaText : "詳しくはこちら"
        }

        // クリックイベント
        nativeAd.setTapEvent(self, handler: nil)
    }
}
